### PR TITLE
Fallback to web flow in case Netcetera 3DS SDK fails

### PR DESCRIPTION
**Pull request by vault087**
AuthorizingPayment is used as a new entry point for authenticated payments. 

AuthorizingPayment tries to use Netcetera SDK and fallback to WebView if any error occurs except:
    - ARes from 3DS Microservice returns "failed" or unknown status
    - ChallengeStatusReceiver get one of these callbacks: cancelled, timedout, completed

AuthPayment is not falling back to webview for Errors:
- NetceteraThreeDSController.Errors.incomplete
- NetceteraThreeDSController.Errors.cancelled
- NetceteraThreeDSController.Errors.timedout
- NetceteraThreeDSController.Errors.authResStatusFailed
- NetceteraThreeDSController.Errors.authResStatusUnknown

image

### CODE REVIEW

File: ExampleApp/Swift/ProductDetailViewController.swift [hunk 1]
```diff
@@ -124,29 +124,30 @@ class ProductDetailViewController: OMSBaseViewController {
...
-        present(alertController, animated: true, completion: nil)
-    }
+            AuthorizingPayment.shared.presentAuthPaymentController(
+                from: self,
+                url: url,
+                expectedWebViewReturnURL: expectedWebReturnURL,
+                deeplinkURL: deeplinkURL,
+                uiCustomization: self.threeDSUICustomization) { [weak self] result in
+                    print("AuthorizingPayment complete with result: \(result)")
+                    print("-")
```
Comment (aashishgurung) [inline ExampleApp/Swift/ProductDetailViewController.swift:138]: Do we need to log these or we forgot to remove?

Comment (vault087) [reply ExampleApp/Swift/ProductDetailViewController.swift]: Example app normally print to console data comes from OmiseSDK, but as I added popup window with status I removed these lines with latest commit.

**Review by AnasNaouchi — COMMENTED** [review]
@vault087  Not sure if we can add tests to this change as I don't see any tests added in this PR and I would say this change is critical because if any mistake arises it will lead to double opening of the auth url.

File: OmiseSDK/3DS/NetceteraThreeDSController.swift [hunk 2]
```diff
@@ -231,43 +281,41 @@ MIIDbzCCAlegAwIBAgIJANp1aztdBEjBMA0GCSqGSIb3DQEBCwUAME4xCzAJBgNVBAMMAmNhMQ4wDAYD
...
-            } catch {
-                print("doChallenge error: \(error)")
-                print("-")
-            }
-//        }
+        do {
+            try transaction.doChallenge(
+                challengeParameters: challengeParameters,
+                challengeStatusReceiver: receiver,
+                timeOut: 5,
```
Comment (danfowler) [inline OmiseSDK/3DS/NetceteraThreeDSController.swift:288]: Where does `5` come from?

Comment (vault087) [reply OmiseSDK/3DS/NetceteraThreeDSController.swift:311]: Timeout is from Netcetera example - "The minimum timeout interval is defined to be 5 minutes" https://3dss.netcetera.com/3dssdk/doc/2.10.0/ios-sdk-api

File: OmiseSDK/3DS/NetceteraThreeDSController.swift [hunk 3]
```diff
@@ -105,7 +156,7 @@ MIIDbzCCAlegAwIBAgIJANp1aztdBEjBMA0GCSqGSIb3DQEBCwUAME4xCzAJBgNVBAMMAmNhMQ4wDAYD
         let configBuilder = ConfigurationBuilder()
         do {
             try configBuilder.add(self.newScheme())
-            try configBuilder.license(key: self.licenceKey)
+            try configBuilder.license(key: self.trialLicenceKey)
```
[the remark this answers is not in the document]
Comment (vault087) [reply OmiseSDK/3DS/NetceteraThreeDSController.swift]: `ConfigurationBuilder` and `license(key:)` function is a part of Netcetera SDK, can't do any changes there but I can rename variable to match the function name

File: OmiseSDK/3DS/NetceteraThreeDSController.swift [hunk 4]
```diff
@@ -61,29 +90,51 @@ MIIDbzCCAlegAwIBAgIJANp1aztdBEjBMA0GCSqGSIb3DQEBCwUAME4xCzAJBgNVBAMMAmNhMQ4wDAYD
...
+                authorizeUrl: authorizeUrl) { response in
+                    guard let response = response else {
+                        onComplete(.failure(NetceteraThreeDSController.Errors.invalidAuthResponse))
+                        return
+                    }
+
+
+                    switch response.status {
+                    case .success:
+                        onComplete(.success(()))
```
[the remark this answers is not in the document]
Comment (vault087) [reply OmiseSDK/3DS/NetceteraThreeDSController.swift:190]: It's an Void success state for defined type Result<Void, Error>, it's a .success(Void) but syntax requires to use ()

**Review by AnasNaouchi — APPROVED** [review]

**Review by danfowler — APPROVED** [review]

**Pull request merged**